### DISCUSSION TITLE
Need help understanding the announce URL

### DISCUSSION THREAD

**Discussion by aarondvail** [torrust/torrust-tracker#705] [Q&A]
Hello

First let me apologize for being a noob.

I need assistance with the announce URL for my private torrust tracker. 

Here's a bullet list of what I'm trying to accomplish in a bash script.
Scan through organized folders (category/author/book as an example) {A simple bash script}
Create a torrent file per book {Using mktorrent}
Add created torrent file to private tracker (Using the post from [How to use private tracker mode?](https://github.com/torrust/torrust-tracker/discussions/476))

And here's where I need help... identifying the different components of the URL: "http://0.0.0.0:7070/announce/U7K9EiczJp5uA8kUavaSMd63Jp4OaSxQ?info_hash=%81%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00&peer_addr=2.137.87.41&downloaded=0&uploaded=0&peer_id=-qB00000000000000001&port=17548&left=0&event=completed&compact=0" 

http://0.0.0.0:7070/announce/ - IP/Port and announce to the tracker.
U7K9EiczJp5uA8kUavaSMd63Jp4OaSxQ - is the key as described in the aforementioned article. I use jq to parse the output to get just the key.
info_hash=%81%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00 - I use sha1sum on the .torrent file to get this value, I'm not sure if that is the correct hash or not...
peer_addr=2.137.87.41 - I'm assuming this can be a bogus address
downloaded=0 - I'm assuming 0 is the desired number here
uploaded=0 - I'm assuming 0 is the desired number here as well
peer_id=-qB00000000000000001 - I have no clue what this is or what the value should be
port=17548 - I'm assuming this can be spoofed
left=0 - I have no idea
event=completed - are we sure?
compact=0 - again no clue

I'm hoping that I can get help on the ones I don't know and possible confirmation of my assumptions. 

Once I get this and one other issue resolved, I'll right up a "how-to" if anyone wants it.

**Comment by josecelano** [comment]
Hi @aarondvail you don't need to apologize. We like having feedback from all users. In fact, yours it's a use case that I think it's very common. I've been thinking about making that process easier.

### Response

The params in the announce request (for the HTTP tracker) are described here:

https://docs.rs/torrust-tracker/3.0.0-alpha.11/torrust_tracker/servers/http/index.html#announce

and officially here:

https://www.bittorrent.org/beps/bep_0003.html

There are only three mandatory fields:

- info_hash
- peer_id
- port

> http://0.0.0.0:7070/announce/ - IP/Port and announce to the tracker.

Corrent.

> U7K9EiczJp5uA8kUavaSMd63Jp4OaSxQ - is the key as described in the aforementioned article. I use jq to parse the output to get just the key.

Correct.

> info_hash=%81%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00%00 - I use sha1sum on the .torrent file to get this value, I'm not sure if that is the correct hash or not...

It's the sha1 of only the `info` dictionary not the whole torrent file. See https://docs.rs/torrust-tracker/3.0.0-alpha.11/torrust_tracker/shared/bit_torrent/info_hash/index.html

> peer_addr=2.137.87.41 - I'm assuming this can be a bogus address

Currently not used. The tracker take the client's IP from the connection info. In the future, it could be a config option (to use this provided in the URL).

> downloaded=0 - I'm assuming 0 is the desired number here
> uploaded=0 - I'm assuming 0 is the desired number here as well
> left=0 - I have no idea

In a BitTorrent announce request, the fields "uploaded", "downloaded", and "left" play crucial roles in communicating the status of a peer within the BitTorrent swarm. Here's what each of these fields represent:

1. **Uploaded**: This field represents the total amount of data that the peer has successfully uploaded to other peers in the swarm, expressed in bytes. It is encoded in base ten ASCII format. This information is important for the BitTorrent protocol as it helps in maintaining a balance between upload and download ratios, encouraging peers to share data actively.

2. **Downloaded**: Similar to "uploaded", this field indicates the total amount of data that the peer has downloaded from other peers in the swarm, also in bytes and encoded in base ten ASCII. This is used to track how much of the file the peer has received and can also influence the peer's priority in the swarm for receiving further data (as BitTorrent protocols often favor peers who contribute more).

3. **Left**: This field signifies the amount of data that the peer still needs to download to complete the file or set of files being shared. It's again in bytes and encoded in base ten ASCII. The "left" field cannot be simply deduced from the "downloaded" field and the total file size because:
   - The peer might be resuming a previous download, and thus may not need to download the entire file.
   - Some parts of the downloaded data might have failed an integrity check (hash check), requiring those parts to be downloaded again. 

These fields are essential for the BitTorrent tracker to manage the swarm effectively. The tracker uses this information to decide which peers to connect for optimal distribution of the file, and also to keep track of overall swarm health and file distribution status.

> peer_id=-qB00000000000000001 - I have no clue what this is or what the value should be

It's the client's unique identifier. You can use any. The prefix is an abbreviation for the software you are using.

> port=17548 - I'm assuming this can be spoofed

Spoofing the port in a BitTorrent context or in general networking can refer to sending packets with a forged source port number. While technically feasible in certain scenarios, it's important to understand the implications and limitations:

1. **Technical Feasibility**: It is technically possible to forge the source port number in IP packets. This is often done in various types of network attacks, such as IP spoofing or certain types of Denial-of-Service (DoS) attacks. In these cases, the attacker sends packets with a forged source IP address and/or port number to hide their identity or to impersonate another system.

2. **Limitations in Two-Way Communication**: In a typical TCP/IP communication (like BitTorrent), a handshake process is required to establish a connection. If a source port is spoofed, any return packets from the target will be directed to the forged port number, potentially at a different host. This makes establishing a legitimate two-way communication channel difficult when source port spoofing is involved.

3. **BitTorrent Context**: In the context of BitTorrent, while you could technically spoof the source port in announce requests or other communications, it wouldn't be very beneficial. BitTorrent relies on establishing connections between peers for file transfers. If a peer spoofs its port, it won't receive incoming connections from other peers, making it ineffective for participating in the swarm.

4. **Detection and Blocking**: Many modern network systems and firewalls are equipped to detect and block spoofed packets. Internet Service Providers (ISPs) may also implement filters to prevent IP spoofing, which includes port spoofing as part of the IP packet header.

5. **Legal and Ethical Considerations**: Spoofing ports, especially for malicious purposes, is illegal in many jurisdictions and violates the ethical standards of most internet communities. It's often associated with harmful activities such as hacking or unauthorized access.

In summary, while port spoofing is technically possible, it has significant limitations in the context of two-way communication protocols like BitTorrent and is generally used for malicious purposes. Additionally, it's often detectable and can lead to legal consequences.

> event=completed - are we sure?

I suppose you want to seed those torrents. Since you have the complete file this should be `completed` so other peers can start downloading from your client.

> compact=0 - again no clue

There are two types of response: normal and compact:

Normal:

```json
{
    "complete": 333,
    "incomplete": 444,
    "interval": 111,
    "min interval": 222,
    "peers": [
       {
          "ip": "105.105.105.105",
          "peer id": "-qB00000000000000001",
          "port": 28784
       },
       {
          "ip": "6969:6969:6969:6969:6969:6969:6969:6969",
          "peer id": "-qB00000000000000002",
          "port": 28784
       }
    ]
}
```

Compact:

```json
{
    "complete": 333,
    "incomplete": 444,
    "interval": 111,
    "min interval": 222,
    "peers": "iiiipp",
    "peers6": "iiiiiiiiiiiiiiiipp"
}
```

See https://docs.rs/torrust-tracker/3.0.0-alpha.11/torrust_tracker/servers/http/index.html#announce

### Other considerations

1. If you want to seed those torrents maybe you don't need to announce them manually. You can add them to a BitTorrent client. The client will announce them to the tracker. You only need to add the tracker URL to the torrent when you build the torrent.
2. I've built a console client recently to make announce requests but only for testing. See https://github.com/torrust/torrust-tracker/blob/develop/src/console/clients/http/app.rs. 
3. If you decide to write a "how-to" you are very welcome to open a PR to our [website](https://github.com/torrust/torrust-website).
4. If you explain your use case I can give you more feedback. I also miss an easy way to share some files, maybe just executing a console command like your bach script.  The problem is even if we make that process easier you have to add those torrents manually to your client. If you use the client I think it does not make sense to announce the torrent manually. Am I missing something?
5. I've been using this command to build torrents: https://github.com/casey/intermodal in case you have been looking for alternatives to mktorrent (nothing against mktorrent just FYI).

**Reply by aarondvail** [reply]
my use case is 2 parts, 
I and a number of my friends use the ARR suite (Sonarr, Radarr, Lidarr, Readarr, and Prowlarr)
     Prowlarr - maintains the index sources for the others
     Sonarr, Radarr, Lidarr and Readarr are automated index scanners and retrievers.

Case 1:
In some cases, I have downloaded files from other sources. and would like to assist the Arr suite in picking up these files. (i.e. a torrent for all books written by Tacitus, which will need to be broken up into individual book torrents)

Case 2:
Share my library(s) with friends.

The goal is to have it as automated as possible, where it creates the torrent, gets a 2 minute access key and announces it. And have it rescan periodically for anything new. And with the indexer attached to Prowlarr (I haven't figured that one out yet, as it's giving a 404 with a URL format: http://localhost:3001/api?t=caps&apikey=MyAccessKey)

For other consideration #4 I think this is taken care of by Prowlarr... I think.... But I've also been looking for info on using a curl command with my transmissionbt instance to trigger the announce (which I too would script :D )

P.S. mktorrent works (I tested it with torrust-tracker as public and I could announce the file manually). It's simple and I built a container for it, so I'm happy with it.
